In [1]:
from sklearn.metrics import log_loss
import numpy as np

#### Pregunta 8

In [2]:
h1 = 2.9557335 + 0.5 * -3.706423 + 1 * -3.6671724

In [3]:
out = 1/(1+np.exp(-h1))

In [4]:
out

0.07144840278053265

#### Pregunta 9

In [5]:
# Suponiendo que clasifica bien
print(log_loss([True],[out],labels=[True,False]))

# Suponiendo que clasifica mal
print(log_loss([False],[out],labels=[True,False]))

2.6387797292223563
0.07412932922235609


#### Pregunta 10

In [12]:
coef_matrix = np.array([
[ 1.2032 , 0.34239826, -0.3272417 ],
[ 1.4500312 , 0.54530275, -0.64213556]
])
bias = np.array([-1.7683319, 1.8029612, -1.7188473])
obs = np.array([0,2])

In [19]:
h = coef_matrix.T @ obs + bias

In [25]:
# Aplico Softmax
probs = np.exp(h)/np.exp(h).sum()
probs

array([0.14621753, 0.85144224, 0.00234022])

#### Cálculo de la entropía cruzada suponiendo que clasifició bien.

En este caso, si se clasificó bien, significa que la clase con mayor probabilidad (clase 1, con 0.851444224) es la correcta. 

Y, considerando que la categorical cross entropy se define como:

$$categorical\_xent=-\sum_{i=1}^N \sum_{l=1}^K d_{i,l}.\log P(y_i=l|X_i)=-\sum_{i=1}^N \sum_{l=1}^K d_{i,l}.\log \frac{\exp(h_l)}{\sum_{m=0}^K{\exp(h_m)}}$$

Para este caso,  $$d_{i,l} = [0,1,0] $$. Por lo que los términos correspondientes a las clases 0 y 2 serán = 0, quedando:

$$categorical\_xent=- 1 * \log \frac{\exp(h_1)}{\sum_{m=0}^K{\exp(h_m)}} = - 1 * \log(0.85144224)$$

In [30]:
log_loss([True],[probs[1]],labels=[True,False])

0.16082361111919713

In [27]:
-np.log(probs[1])

0.16082361111919713

Si en cambio la clasificación correcta hubiera sido la primer clase (clase 0)...

In [31]:
-np.log(probs[0])

1.9226598111191973

#### Problema 11

In [32]:
coefs = np.array([
[-10.021508 , 9.780544 , -0.3323118 , 0.35372728],
[ 0.13239847, -0.1332541 , 8.324003 , -8.686057 ]
])

intercept = np.array([ 0.19760127, -0.18874079, 0.03835499, -0.04586953])

obs = np.array([0.5,0.5])

In [34]:
h = obs.dot(coefs) + intercept
print(h)

[-4.7469535   4.63490416  4.03420059 -4.21203439]


In [39]:
probs = 1/(1+np.exp(-h))

In [40]:
probs

array([0.00860343, 0.99038628, 0.98260801, 0.01459988])

In [41]:
log_loss([False,True,True,False],probs,labels=[True,False])

0.012638348909066743